### PR TITLE
Feature 4893 performance improvements

### PR DESCRIPTION
**Pull request by AndreasSundebo**
#### Description

Using the VS2019 Performance Profiler several places were identified where the performance of LEAN could be improved by applying minor tweaks for a respectable total performance improvement of 10-20% during backtesting.

I was limited to profiling and runtime tests as I haven't been able to get Python working locally with LEAN, but all tests passed in the build pipeline.

For ease of reviewing each commit in the PR covers its own minor tweak.

#### Related Issue

#4893 
#### Motivation and Context

Performance
#### Requires Documentation Change

No
#### How Has This Been Tested?

VS2019 Performance profiler was run before and after to verify that bottle necks had been removed or reduced. Backtesting runtimes were verified lower total runtime while providing the same stats as before.

#### Performance tests
BasicTemplateBenchmark C# using EURUSD from 2007-04-01 to 2017-01-01 :
master
25,48 seconds at 142k data points per second. Processing total of 3.627.182 data points.
24,68 seconds at 147k data points per second. Processing total of 3.627.182 data points.
25,47 seconds at 142k data points per second. Processing total of 3.627.182 data points.
PR (measurements based on PR as of 2020-11-03)
23,07 seconds at 157k data points per second. Processing total of 3.627.182 data points.
23,47 seconds at 155k data points per second. Processing total of 3.627.182 data points.
22,88 seconds at 159k data points per second. Processing total of 3.627.182 data points.

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which improves implementation)
- [X] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:

- [X] My code follows the code style of this project.
- [X] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [ ] I have added tests to cover my changes.
- [X] All new and existing tests passed.
- [X] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

**Comment by jaredbroad** [comment]
Welcome to LEAN @AndreasSundebo, thank you for the PR. We'll review and post feedback or tweaks ASAP.

**Comment by AndreasSundebo** [comment]
@Martin-Molinero @jaredbroad Added performance tests as requested.

File: Common/Util/Composer.cs [hunk 1]
```diff
@@ -132,7 +132,7 @@ public Composer()
 
         private CompositionContainer _compositionContainer;
         private readonly List<Type> _exportedTypes = new List<Type>();
-        private readonly Task<List<ComposablePartDefinition>> _composableParts;
+        private readonly Task<IEnumerable<ComposablePartDefinition>> _composableParts;
```
Comment (Martin-Molinero) [inline Common/Util/Composer.cs:135]: Interesting, returning a list in `master` has the benefit that it's performed in _parallel_ in the task were returning an IEnumerable will deffer the operation to **each** request _synchronously_ which will delay requests to the composer

File: Engine/AlgorithmManager.cs [hunk 2]
```diff
@@ -512,7 +513,7 @@ public void Run(AlgorithmNodePacket job, IAlgorithm algorithm, ISynchronizer syn
                                 }
 
                                 // scan for time after we've pumped all the data through for this consolidator
-                                consolidator.Scan(timeKeeper.GetLocalTimeKeeper(update.Target.ExchangeTimeZone).LocalTime);
+                                consolidator.Scan(localTime);
```
Comment (Martin-Molinero) [inline Engine/AlgorithmManager.cs:516]: Another nice one 💯

File: Engine/AlgorithmManager.cs [hunk 3]
```diff
@@ -586,16 +587,6 @@ public void Run(AlgorithmNodePacket job, IAlgorithm algorithm, ISynchronizer syn
...
-
-                    // TODO: For backwards compatibility only. Remove in 2017
-                    // For compatibility with Forex Trade data, moving
-                    if (timeSlice.Slice.QuoteBars.Count > 0)
-                    {
-                        foreach (var tradeBar in timeSlice.Slice.QuoteBars.Where(x => x.Key.ID.SecurityType == SecurityType.Forex))
-                        {
-                            timeSlice.Slice.Bars.Add(tradeBar.Value.Collapse());
-                        }
-                    }
```
Comment (Martin-Molinero) [inline Engine/AlgorithmManager.cs:598]: Cool! We should review for more `TODO` in `master` 😅

**Review by Martin-Molinero — CHANGES_REQUESTED** [review]
Thanks for the PR @AndreasSundebo!
Some nice improvements, leaving some requests & concerns

File: Brokerages/Backtesting/BacktestingBrokerage.cs [hunk 4]
```diff
@@ -251,7 +251,7 @@ public virtual void Scan()
                 var stillNeedsScan = false;
 
                 // process each pending order to produce fills/fire events
-                foreach (var kvp in _pending.OrderBy(x => x.Key))
+                foreach (var kvp in _pending)
```
[the remark this answers is not in the document]
Comment (AndreasSundebo) [reply Brokerages/Backtesting/BacktestingBrokerage.cs]: OK, I've reverted this change.

File: Common/Util/Composer.cs (hunk 1 above)
Comment (AndreasSundebo) [reply Common/Util/Composer.cs]: That's certainly not intended. I'm reverting this.

File: Common/Interfaces/ISubscriptionDataConfigProvider.cs [hunk 5]
```diff
@@ -28,6 +28,6 @@ public interface ISubscriptionDataConfigProvider
         /// Gets a list of all registered <see cref="SubscriptionDataConfig"/> for a given <see cref="Symbol"/>
         /// </summary>
         /// <remarks>Will not return internal subscriptions by default</remarks>
-        List<SubscriptionDataConfig> GetSubscriptionDataConfigs(Symbol symbol, bool includeInternalConfigs = false);
+        IEnumerable<SubscriptionDataConfig> GetSubscriptionDataConfigs(Symbol symbol, bool includeInternalConfigs = false);
```
[the remark this answers is not in the document]
Comment (AndreasSundebo) [reply Common/Interfaces/ISubscriptionDataConfigProvider.cs]: I'm reverting all List -> IEnumerable changes for the time being.

File: Algorithm/QCAlgorithm.Indicators.cs [hunk 6]
```diff
@@ -2162,7 +2162,7 @@ private void WarmUpIndicatorImpl<T>(Symbol symbol, TimeSpan period, Action<T> ha
             where T : class, IBaseData
         {
             IDataConsolidator consolidator;
-            if (SubscriptionManager.SubscriptionDataConfigService.GetSubscriptionDataConfigs(symbol).Count > 0)
+            if (SubscriptionManager.SubscriptionDataConfigService.GetSubscriptionDataConfigs(symbol).Any())
```
[the remark this answers is not in the document]
Comment (AndreasSundebo) [reply Algorithm/QCAlgorithm.Indicators.cs]: Reverting as this change is no longer needed due to reverting the List -> IEnumerable changes.

**Review by Martin-Molinero — APPROVED** [review]

**Pull request merged**